packages/stylelint-plugin/src/rules/no-invalid-use-of-tokens.ts

```diff
@@ -0,0 +1,354 @@
+  return deprecatedKaizenTokenPattern.test(value)
+}
+
+// Most CSS standard functions are allowed to contain CSS variables.
+const functionsThatSupportCSSVariables = new Set(cssStandardFunctions)
+// SASS overloads the "saturate" function and computes it at compile time.
+// We therefore don't allow it to be used with CSS variable tokens.
+functionsThatSupportCSSVariables.delete("saturate")
+
+/**
+ * This is the set of functions that we know don't support CSS variables.
+ * Rather than have an allow-list of functions that we know do support CSS variables, a deny-list should
+ * be a lot less strict on users.
+ * It should say "we know that AT LEAST these functions won't support CSS variables"
+ * rather than "you can only use CSS variables within these functions".
+ * The former allows for use of CSS variables in functions we've never seen before, such as user defined ones like `ca-margin`.
+ *
+ * At the time of writing this, we've just populated this list with color manipulation functions, then subtracted ones
+ * from the allow list so that any potential overloaded functions (ones that work both with the SASS compiler and natively in the browser, like rgba()) are allowed.
+ * https://www.tutorialsteacher.com/sass/sass-color-functions
+ */
+const functionsThatDontSupportCSSVariables = new Set(
+  [
+    "mix",
+    "shade",
+    "tint",
+    "darken",
+    "lighten",
+    "adjust-hue",
+    "saturate",
+    "desaturate",
+    "add-alpha",
+    "transparentize",
+    "adjust-color",
+    "scale-color",
+    "change-color",
+    "opacify",
+  ].filter(i => !functionsThatSupportCSSVariables.has(i))
+)
+
+// The AtRules which shouldn't contain CSS variables in their parameters
+// https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule
+const atRulesThatDontSupportCSSVars = new Set([
+  "media",
+  "supports",
+  "document",
+  "page",
+  "font-face",
+  "keyframes",
+  "viewport",
+  "counter-style",
+  "font-feature-values",
+  "property",
+  "color-profile",
+])
+
+/**
+ * This function attempts to detect and autofix invalid usages of tokens, including ones which are deprecated or used in an incorrect way within functions.
+ * If the linter determines it can't automatically do a replacement, it will report an unfixable declaration (perhaps the replacement would cause an error).
+ *
+ * Note: There is definitely some weirdness here. In order to clean it up, we really need a better value parser, in particular one that has a better hierarchical structure (one which we can go up and down between parents and children) and a better understanding of SASS and LESS constructs like negation and string interpolation.
+ *
+ * @returns a new string value if there we're changes, or undefined if there weren't any
+ */
+function detectAndFixInvalidTokens(
+  nodeValue: string,
+  postcssNode: Declaration | AtRule,
+  options: Options
+): string | undefined {
+  let parsedValue = postcssValueParser(nodeValue)
+  let newValue = nodeValue
```

Comment:
I wonder if `currentValue: string` is better than newValue, because it's not necessarily different? Hard to say